source/docs/guide/src/nonlinear_bitvec.md

```diff
@@ -0,0 +1,287 @@
+
+```rust
+pub proof fn lemma_mod_difference_equal_helper(x: int, y:int, d:int, small_x:int, small_y:int, tmp1:int, tmp2:int) by(integer_ring)
+    requires
+        small_x == x % d,
+        small_y == y % d,
+        tmp1 == (small_y - small_x) % d,
+        tmp2 == (y - x) % d,
+    ensures
+        (tmp1 - tmp2) % d == 0
+{}
+pub proof fn lemma_mod_difference_equal(x: int, y: int, d: int) by(nonlinear_arith)
+    requires
+        d > 0,
+        x <= y,
+        x % d <= y % d,
+        y - x < d
+    ensures
+        y % d - x % d == y - x
+{
+    let small_x = x % d;
+    let small_y = y % d;
+    let tmp1 = (small_y - small_x) % d;
+    let tmp2 = (y - x) % d;
+    lemma_mod_difference_equal_helper(x,y,d, small_x, small_y, tmp1, tmp2);
+}
+```
+
+In the below `lemma_mod_between`, we want to prove that `x % d <= z % d < y % d`. However, `integer_ring` only supports equalities, so we need cannot prove `lemma_mod_between` directly. Instead, we provide facts that can help assist the proof. The helper lemma provides 1) `x % d - y % d == x - y  (mod d)` and 2) ` y % d - z % d == y - z  (mod d)`. The rest of the proof is done by `by(nonlinear_arith)`.
+
+```rust
+pub proof fn lemma_mod_between_helper(x: int, y: int, d: int, small_x:int, small_y:int, tmp1:int) by(integer_ring)
+    requires
+        small_x == x % d,
+        small_y == y % d,
+        tmp1 == (small_x - small_y) % d,
+    ensures
+        (tmp1 - (x-y)) % d == 0
+{}
+
+// note that below two facts are from the helper function, and the rest are done by `by(nonlinear_arith)`.
+// x % d - y % d == x - y  (mod d)
+// y % d - z % d == y - z  (mod d)
+pub proof fn lemma_mod_between(d: int, x: int, y: int, z: int) by(nonlinear_arith)
+    requires
+        d > 0,
+        x % d < y % d,
+        y - x <= d,
+        x <= z < y
+    ensures
+        x % d <= z % d < y % d
+{
+    let small_x = x % d;
+    let small_y = y % d;
+    let small_z = z % d;
+    let tmp1 = (small_x - small_z) % d;
+    lemma_mod_between_helper(x,z,d, small_x, small_z, tmp1);
+
+    let tmp2 = (small_z - small_y) % d;
+    lemma_mod_between_helper(z,y,d, small_z, small_y, tmp2);    
+}
+```
+
+
+#### Example 2: Proving properties on bounded integers with the help of `integer_ring`
+
+Since `integer_ring` proofs only support `int`, you need to include explicit bounds when you want to prove properties about bounded integers. For example, in order to use the proof `lemma_mod_after_mul` on `u32`s, `lemma_mod_after_mul_u32` must ensure that all arguments are within the proper bounds before passing them to `lemma_mod_after_mul`.  
+
+If a necessary bound (e.g., `m > 0`) is not included, Verus will fail to verify the proof.
+
```

Comment:
I'm not sure I understand this example.  It seems like `lemma_mod_after_mul` is a generic property, and `lemma_mod_after_mul_32` is just a specialization?  I'm not sure what lesson is being taught.